api/explorer/api/ingest.py

```diff
@@ -101,89 +101,100 @@ def _edge_from_build_input_type(build_input_type: model.BuildInputType) -> graph
     """
-    Turn a parsed NixGraph from core model into two lists, a list of nodes (Package)
-    and a list of edges, in the API model.
+    Processes nix_package, returning the new edges that must be created,
+    and augmenting ctxt with both the new packages (ctxt.packages) and
+    the edges to add later on (ctxt.pending_edges)
 
     Args:
-        nix_graph (model.NixGraph): parsed NixGraph (core model)
-
+        nix_package (model.Package): parsed Nix package (core model)
+        ctxt (_IngestionContext): accumulator of the ingestion
     Returns:
         tuple[
-            List[graph.Package],
-            List[tuple[graph.Package, graph.Edge, graph.Package]]
-        ]:
-        A tuple containing the list of nodes (packages) and the list of edges,
-    both in the graph API model.
-
-    Raises:
-        Exception: when the output path of an edge's vertex does not correspond to
-    any derivation in the NixGraph
+            list[graph.Package],
+            list[tuple[graph.Package, graph.Edge, graph.Package]]:
+        ]
+        The packages and edges created while processing nix_package. In the edges,
+        a tuple (A, edge, B) means A -edge-> B.
     """
-    # Map from output path to package node
-    packages: dict[str, graph.Package] = {}
-
-    # A dictionary that maps pairs of package output paths to edge objects.
-    edge_refs = {}
-
-    # A list of tuples containing package, edge, and dependency package objects.
-    edges = []
-
-    # Iterate over each package in the Nix graph
-    for p in nix_graph.packages:
-        parsed = safe_parse_package(p)
-        for ps in parsed:
-            packages[ps.outputPath] = ps
-            # Populate edge_refs dictionary with package output path pairs
-            # and their corresponding edge objects
-            edge_refs |= {
-                (
-                    ps.outputPath,
-                    build_input.output_path,
-                ): _edge_from_build_input_type(build_input.build_input_type)
-                for build_input in p.build_inputs
-            }
-    missing = []
-    for paths, edge in edge_refs.items():
-        package_output_path, dependency_output_path = paths
-        if package_output_path not in packages:
-            raise Exception(
-                "Package metadata could not be found for the output path "
-                "corresponding to the edge's input vertex. "
-                "This should be impossible! "
-                f"Output path: {package_output_path} "
-            )
-        if dependency_output_path not in packages:
-            missing.append(dependency_output_path)
+    edges: list[tuple[graph.Package, graph.Edge, graph.Package]] = []
+
+    parsed_pkgs: list[graph.Package] = safe_parse_package(nix_package)
+    for pkg in parsed_pkgs:
+        # TODO: do we have the invariant that packages[ps.outputPath]
+        # is not mapped here? If yes should we check it?
```

Comment:
In Nix, a derivation outPath is "supposed" to be a unique ID for a derivation thanks to the hash it includes.

Reply:
OK I see, I'm not adding the assertion then, since it's something external we have no control over. So it doesn't make much sense to crash over it.

Reply:
I removed the comment, hence resolving.